cli: reana-client validate fails when not server_url or token is set

https://travis-ci.org/diegodelemos/reana-demo-atlas-recast/builds/413629487#L582
Output from a Travis build on a REANA Demo:
```console
$ reana-client validate | grep -q ' is a valid '
[ERROR] REANA client is not connected to any REANA cluster.
Please set REANA_SERVER_URL environment variable to the remote REANA cluster you would like to connect to.
For example: export REANA_SERVER_URL=https://reana.cern.ch/
```
`reana-client validate` should work even if there is no `REANA_SERVER_URL` nor `REANA_ACCESS_TOKEN`.

--- a/reana_client/cli/__init__.py
+++ b/reana_client/cli/__init__.py
@@ -41,20 +41,7 @@ class Config(object):
 
     def __init__(self):
         """Initialize config variables."""
-        server_url = os.environ.get('REANA_SERVER_URL', None)
-
-        if not server_url:
-            logging.error(
-                'REANA client is not connected to any REANA cluster.\n'
-                'Please set REANA_SERVER_URL environment variable to '
-                'the remote REANA cluster you would like to connect to.\n'
-                'For example: export REANA_SERVER_URL=https://reana.cern.ch/')
-            sys.exit(1)
-
-        logging.info('REANA Server URL ($REANA_SERVER_URL) is: {}'
-                     .format(server_url))
-
-        self.client = Client(server_url)
+        self.client = None
 
 
 @click.group()

--- a/reana_client/cli/files.py
+++ b/reana_client/cli/files.py
@@ -32,6 +32,7 @@ import tablib
 
 from ..config import ERROR_MESSAGES, default_user
 from ..errors import FileUploadError
+from reana_client.decorators import with_api_client
 from reana_commons.utils import click_table_printer
 
 
@@ -69,6 +70,7 @@ def files(ctx):
     default=os.environ.get('REANA_ACCESS_TOKEN', None),
     help='Access token of the current user.')
 @click.pass_context
+@with_api_client
 def get_files(ctx, workflow, _filter,
               output_format, access_token):
     """List workflow workspace files."""
@@ -146,6 +148,7 @@ def get_files(ctx, workflow, _filter,
     default=os.environ.get('REANA_ACCESS_TOKEN', None),
     help='Access token of the current user.')
 @click.pass_context
+@with_api_client
 def download_files(ctx, workflow, file_, output_directory, access_token):
     """Download workflow workspace file(s)."""
     logging.debug('command: {}'.format(ctx.command_path.replace(" ", ".")))
@@ -221,6 +224,7 @@ def download_files(ctx, workflow, file_, output_directory, access_token):
     default=os.environ.get('REANA_ACCESS_TOKEN', None),
     help='Access token of the current user.')
 @click.pass_context
+@with_api_client
 def upload_files(ctx, workflow, filenames, access_token):
     """Upload file(s) to workflow workspace."""
     logging.debug('command: {}'.format(ctx.command_path.replace(" ", ".")))

--- a/reana_client/cli/ping.py
+++ b/reana_client/cli/ping.py
@@ -26,9 +26,12 @@ import traceback
 
 import click
 
+from reana_client.decorators import with_api_client
 
-@click.command()
+
+@click.command('ping')
 @click.pass_context
+@with_api_client
 def ping(ctx):
     """Health check REANA Server."""
     try:

--- a/reana_client/cli/workflow.py
+++ b/reana_client/cli/workflow.py
@@ -28,14 +28,17 @@ import traceback
 from enum import Enum
 
 import click
-from jsonschema.exceptions import ValidationError
 import tablib
 import yaml
+from jsonschema.exceptions import ValidationError
+from reana_commons.utils import click_table_printer
+
+from reana_client.decorators import with_api_client
 
 from ..config import ERROR_MESSAGES, reana_yaml_default_file_path
-from ..utils import get_workflow_name_and_run_number, load_reana_spec, \
-    load_workflow_spec, is_uuid_v4, workflow_uuid_or_name
-from reana_commons.utils import click_table_printer
+from ..utils import (get_workflow_name_and_run_number, is_uuid_v4,
+                     load_reana_spec, load_workflow_spec,
+                     workflow_uuid_or_name)
 
 
 class _WorkflowStatus(Enum):
@@ -78,6 +81,7 @@ def workflow(ctx):
     count=True,
     help='Set status information verbosity.')
 @click.pass_context
+@with_api_client
 def workflow_workflows(ctx, _filter, output_format, access_token,
                        verbose):
     """List all workflows user has."""
@@ -168,6 +172,7 @@ def workflow_workflows(ctx, _filter, output_format, access_token,
     default=os.environ.get('REANA_ACCESS_TOKEN', None),
     help='Access token of the current user.')
 @click.pass_context
+@with_api_client
 def workflow_create(ctx, file, name, skip_validation, access_token):
     """Create a REANA compatible workflow from REANA spec file."""
     logging.debug('command: {}'.format(ctx.command_path.replace(" ", ".")))
@@ -240,6 +245,7 @@ def workflow_create(ctx, file, name, skip_validation, access_token):
     nargs=-1,
 )
 @click.pass_context
+@with_api_client
 def workflow_start(ctx, workflow, access_token, parameters):
     """Start previously created workflow."""
     logging.debug('command: {}'.format(ctx.command_path.replace(" ", ".")))
@@ -313,6 +319,7 @@ def workflow_start(ctx, workflow, access_token, parameters):
     count=True,
     help='Set status information verbosity.')
 @click.pass_context
+@with_api_client
 def workflow_status(ctx, workflow, _filter, output_format,
                     access_token, verbose):
     """Get status of previously created workflow."""
@@ -443,6 +450,7 @@ def workflow_status(ctx, workflow, _filter, output_format,
     default=os.environ.get('REANA_ACCESS_TOKEN', None),
     help='Access token of the current user.')
 @click.pass_context
+@with_api_client
 def workflow_logs(ctx, workflow, access_token):
     """Get workflow logs."""
     logging.debug('command: {}'.format(ctx.command_path.replace(" ", ".")))
@@ -471,7 +479,7 @@ def workflow_logs(ctx, workflow, access_token):
 
 @click.command(
     'validate',
-    help='Get workflow logs.')
+    help='Validate the REANA specification.')
 @click.option(
     '-f',
     '--file',

--- a/reana_client/decorators.py
+++ b/reana_client/decorators.py
@@ -1,0 +1,63 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of REANA.
+# Copyright (C) 2018 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software
+# Foundation; either version 2 of the License, or (at your option) any later
+# version.
+#
+# REANA is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# REANA; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
+# Suite 330, Boston, MA 02111-1307, USA.
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization or
+# submit itself to any jurisdiction.
+"""REANA command line decorators."""
+
+import logging
+import os
+import sys
+
+from click.core import Context
+
+from reana_client.api import Client
+
+
+def with_api_client(f):
+    """Decorator to inject the REANA Server API client to a Click command."""
+    def wrapper(*args, **kwargs):
+        """Initialize config for API client.
+
+        This decorator should be used right after @click.pass_context since it
+        is injecting the API client inside the ``ctx`` object.
+        """
+        # Since we require the @click.pass_context decorator to be passed
+        # before, we know ``ctx`` is the first argument.
+        ctx = args[0]
+        if isinstance(ctx, Context):
+            server_url = os.environ.get('REANA_SERVER_URL', None)
+
+            if not server_url:
+                logging.error(
+                    'REANA client is not connected to any REANA cluster.\n'
+                    'Please set REANA_SERVER_URL environment variable to '
+                    'the remote REANA cluster you would like to connect to.\n'
+                    'For example: export '
+                    'REANA_SERVER_URL=https://reana.cern.ch/')
+                sys.exit(1)
+
+            logging.info('REANA Server URL ($REANA_SERVER_URL) is: {}'
+                         .format(server_url))
+            ctx.obj.client = Client(server_url)
+        else:
+            raise Exception(
+                'This decorator should be used after click.pass_context.')
+        f(*args, **kwargs)
+    return wrapper

